File: apps/desktop/Sources/Dahlia/Models/WorkspaceGenerationSettings.swift
import Foundation

/// Shared generation defaults, stored with the Workspace and captured by each job.
struct WorkspaceGenerationSettings: Codable, Equatable, Sendable {
    enum SummaryMode: String, Codable, CaseIterable, Identifiable, Sendable {
        case local, remote
        var id: Self { self }
    }

    enum Workflow: String, Codable, CaseIterable, Identifiable, Sendable {
        case transcribeThenSummarize, combined
        var id: Self { self }
    }

    struct RemoteProcessing: Codable, Equatable, Sendable {
        var workflow: Workflow = .combined
        var summaryModel: String?
        var reasoningEffort: String?
    }

    struct Processing: Codable, Equatable, Sendable {
        /// Transcription location. Summary routing follows Workspace ownership.
        var location: SummaryMode = .local
        var remote = RemoteProcessing()
    }

    struct Summary: Codable, Equatable, Sendable {
        var style: SummaryStyle = .detailed
        var detailLevel: SummaryDetailLevel { style.detailLevel }
    }

    struct GenerationPreferences: Codable, Equatable, Sendable {
        var processing: Processing
        var summary: Summary
        var outputLanguage: SummaryLanguage
    }

    struct LocalProcessing: Codable, Equatable, Sendable {
        var model = "gpt-5.6-luna"
        var reasoningEffort = "high"
    }

    struct LegacyTranscription: Codable, Equatable, Sendable {
        var localeIdentifier: String
        var automaticLanguageDetection: Bool
        var languageScope: TranscriptionLanguageScope
        var languageIdentifiers: [String]
        var liveTranscriptDraft: Bool
    }

    var processing = Processing()
    var summary = Summary()
    var outputLanguage: SummaryLanguage = .ja
    var local = LocalProcessing()
    var automaticProcessing = true
    private(set) var legacyTranscription: LegacyTranscription?

    init(
        processing: Processing = Processing(),
        summary: Summary = Summary(),
        outputLanguage: SummaryLanguage = .ja,
        local: LocalProcessing = LocalProcessing(),
        automaticProcessing: Bool = true
    ) {
        self.processing = processing
        self.summary = summary
        self.outputLanguage = outputLanguage
        self.local = local
        self.automaticProcessing = automaticProcessing
    }

    private enum CodingKeys: String, CodingKey {
        case processing, summary, outputLanguage, local, automaticProcessing, transcription
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        processing = try values.decodeIfPresent(Processing.self, forKey: .processing) ?? Processing()
        summary = try values.decodeIfPresent(Summary.self, forKey: .summary) ?? Summary()
        outputLanguage = try values.decodeIfPresent(SummaryLanguage.self, forKey: .outputLanguage) ?? .ja
        local = try values.decodeIfPresent(LocalProcessing.self, forKey: .local) ?? LocalProcessing()
        automaticProcessing = try values.decodeIfPresent(Bool.self, forKey: .automaticProcessing) ?? true
        legacyTranscription = try values.decodeIfPresent(LegacyTranscription.self, forKey: .transcription)
    }

    func encode(to encoder: Encoder) throws {
        var values = encoder.container(keyedBy: CodingKeys.self)
        try values.encode(processing, forKey: .processing)
        try values.encode(summary, forKey: .summary)
        try values.encode(outputLanguage, forKey: .outputLanguage)
        try values.encode(local, forKey: .local)
        try values.encode(automaticProcessing, forKey: .automaticProcessing)
    }

    static func == (lhs: Self, rhs: Self) -> Bool {
        lhs.processing == rhs.processing && lhs.summary == rhs.summary && lhs.outputLanguage == rhs.outputLanguage
            && lhs.local == rhs.local && lhs.automaticProcessing == rhs.automaticProcessing
    }

    var generationPreferences: GenerationPreferences {
        .init(processing: processing, summary: summary, outputLanguage: outputLanguage)
    }
}

File: apps/server/src/workspace-generation-settings.ts
import { z } from "zod";

export const outputLanguageSchema = z.enum(["ja", "en", "zh", "ko", "fr", "de", "es"]);
export const summaryModeSchema = z.enum(["local", "remote"]);
export const summaryDetails = ["low", "medium", "high", "xhigh", "max"] as const;
export function normalizeSummaryDetail(value: string): string {
  switch (value) {
    case "concise": return "low";
    case "standard": return "medium";
    case "detailed": return "high";
    case "eventSession": return "xhigh";
    default: return value;
  }
}
export const summaryDetailSchema = z.enum(summaryDetails);
export const summaryModelSettingsSchema = z.object({
  model: z.string().trim().min(1).max(200),
  reasoningEffort: z.enum(["none", "minimal", "low", "medium", "high", "xhigh", "max", "ultra"]),
}).strict();
export const summaryStyles = ["concise", "standard", "detailed", "eventSummary", "eventTimeline"] as const;
export const summaryStyleSchema = z.enum(summaryStyles);
export function summaryStyleDetail(style: z.infer<typeof summaryStyleSchema>): z.infer<typeof summaryDetailSchema> {
  return { concise: "low", standard: "medium", detailed: "high", eventSummary: "xhigh", eventTimeline: "max" }[style] as z.infer<typeof summaryDetailSchema>;
}
const modelPreference = z.string().trim().min(1).max(200);
export const remoteProcessingSchema = z.object({
  workflow: z.enum(["transcribeThenSummarize", "combined"]),
  summaryModel: modelPreference.optional(),
  transcriptionModel: modelPreference.optional(),
  reasoningEffort: summaryModelSettingsSchema.shape.reasoningEffort.optional(),
}).strict();
export const processingSchema = z.object({ location: summaryModeSchema, remote: remoteProcessingSchema }).strict();
const summarySchema = z.object({ style: summaryStyleSchema }).strict();
export const transcriptionSettingsSchema = z.object({
    localeIdentifier: z.string().max(100).regex(/^[A-Za-z]{2,3}(?:[-_][A-Za-z0-9]{2,8})*$/),
    automaticLanguageDetection: z.boolean(),
    languageScope: z.enum(["all", "selected"]),
    languageIdentifiers: z.array(z.string().regex(/^[a-z]{2,3}(?:-[A-Z][a-z]{3})?$/)).max(200),
    liveTranscriptDraft: z.boolean(),
}).strict().refine((value) => !value.automaticLanguageDetection || value.languageScope === "all" || value.languageIdentifiers.length > 0,
  { message: "Select at least one language for automatic detection", path: ["languageIdentifiers"] });
export const generationPreferencesSchema = z.object({ outputLanguage: outputLanguageSchema, processing: processingSchema, summary: summarySchema, transcription: transcriptionSettingsSchema }).strict();

export type GenerationPreferences = z.infer<typeof generationPreferencesSchema>;
export const DEFAULT_GENERATION_PREFERENCES: GenerationPreferences = {
  outputLanguage: "ja",
  processing: { location: "local", remote: { workflow: "transcribeThenSummarize" } },
  summary: { style: "detailed" },
  transcription: { localeIdentifier: "ja-JP", automaticLanguageDetection: false, languageScope: "all", languageIdentifiers: [], liveTranscriptDraft: false },
};
export const workspaceGenerationSettingsSchema = generationPreferencesSchema.extend({
  local: summaryModelSettingsSchema,
  automaticProcessing: z.boolean(),
}).strict();
export type WorkspaceGenerationSettings = z.infer<typeof workspaceGenerationSettingsSchema>;
export const DEFAULT_WORKSPACE_GENERATION_SETTINGS: WorkspaceGenerationSettings = {
  ...DEFAULT_GENERATION_PREFERENCES,
  local: { model: "gpt-5.6-luna", reasoningEffort: "high" },
  automaticProcessing: true,
};
